boards/common/particle-mesh/bootloader.c

```diff
@@ -0,0 +1,93 @@
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @author  Christian Amsüss <chrysn@fsfe.org>
+ */
+
+#ifdef PARTICLE_MONOFIRMWARE
+#include <stdint.h>
+#include "vectors_cortexm.h"
+
+extern const void *_isr_vectors;
+
+/* As described in Particle's dynalib/inc/module_info.h (there under LGPL-3
+ * terms; detailed license compatibility discussion skipped for not crossing
+ * any threshold of originality) */
+typedef struct module_info_t {
+    const void* module_start_address;
+    const void* module_end_address;
+    uint8_t reserved;
+    uint8_t flags;
+    uint16_t module_version;
+    uint16_t platform_id;
+    uint8_t  module_function;
+    uint8_t  module_index;
+    uint32_t dependency; /* was module_dependency_t; kept in here to ensure it's blank */
+    uint32_t dependency2; /* was module_dependency_t */
+} module_info_t;
+
+/* 64 words is the distance between the known ISR table and the fixed-position
+ * module_info at 0x200. If the ISR length is changed for any reason, this must
+ * be adapted until particle_monofirmware_module_info is precisely at 0x30200.
+ * (monofirmware-tool.py will complain if it isn't).
+ *
+ * This is not `static` as *something* from this module needs to be visible to
+ * the outside for the __attribute__((used)) and the KEEP on the section in the
+ * linker script to work; see also the reference to it in board.c. */
+ISR_VECTOR(50) const uint32_t particle_monofirmware_padding[64] = {0, };
+
+#ifdef PARTICLE_MONOFIRMWARE_CHECKSUMLIMIT
+/* Watch the ISR vector sequence here: This is placed *after* the module_info
+ * in the flash (52 > 51), but placed before it in the code to be referencable
+ * from there.
+ *
+ * The actual value is replaced by monofirmware-tool.py.
+ * */
+ISR_VECTOR(52) static const uint8_t particle_monofirmware_checksum[4] = {0, 0, 0, 0};
+#endif
+
+ISR_VECTOR(51) static const struct module_info_t particle_monofirmware_module_info = {
+    .module_start_address = &_isr_vectors,
+#ifdef PARTICLE_MONOFIRMWARE_CHECKSUMLIMIT
+    .module_end_address = &particle_monofirmware_checksum,
+    /* else, it is set by the monofirmware-tool.
+     *
+     * For that (more common) case where we check the full firmware, we would
+     * want to set set
+     *
+     * .module_end_address = &_etext + (&_erelocate - &_srelocate),
+     *
+     * but that can not be determined at compile time. Instead, this will be
+     * populated by monofirmware-tool.py, which needs to write in here anyway
+     * for the checksum. */
+#endif
+    .module_version = 0x138, /* module version that micropython uses */
+
+    /* Platform IDs indicae the actual board: see Particle's build/platform-id.mk */
+#ifdef BOARD_PARTICLE_ARGON
+    .platform_id = 12,
+#endif
+#ifdef BOARD_PARTICLE_BORON
+    .platform_id = 13,
+#endif
+#ifdef BOARD_PARTICLE_XENON
+    .platform_id = 14,
+#endif
+
+    .module_index = 3, /* MOD_FUNC_MONO_FIRMWARE from dynalib/inc/module_info.h */
+};
+
+#else
+typedef int dont_be_pedantic;
+#endif
+
+/** @} */
```

Comment:
Is there a way to signal the bootloader that it should not boot the firmware but stay in DFU mode, i.e. can you provide a `usb_board_reset_in_bootloader()` function?

Reply:
Looks like it; the original code has a HAL with a [`HAL_Core_Enter_Safe_Mode` function](https://github.com/particle-iot/device-os/blob/c21d5b8f39c8a013e19f904800808673f369c6e3/hal/src/nRF52840/core_hal.c#L457). It'll need some expanding, but I think I can do that. (Which is nice given the reset procedure needs to fingers and some timing.).

Reply:
I've picked the parts from the original sources, and as with the firmware addresses it's as weird to use as the module info (unless you're building a firmware that uses all their ecosystem).

What I couldn't figure out though is why this isn't interacting with the baud-rate-600 automatism (which I'm not particularly fond of but haven't contributed anything better either...). I *can* use the command line reset with

```
[ go to bootloader mode using buttons ]
[ it's blinking yellow ]
$ make BOARD=particle-xenon PARTICLE_MONOFIRMWARE=1 -C tests/usbus_cdc_acm_stdio all flash term PORT=/dev/ttyACM0
[...]
> help
Command              Description
---------------------------------------
text                 Generates long text for testing stdio buffer
reboot               Reboot the node
version              Prints current RIOT_VERSION
bootloader           Reboot to bootloader
pm                   interact with layered PM subsystem
ps                   Prints information about running threads.
> bootloader
[ blinks yellow again and accepts uploads ]
```

but just setting baud rate 600 reboots the board without going to the boot loader. Is there any further configuration necessary to enable this?

Reply:
> just setting baud rate 600 reboots the board without going to the boot loader.

That's the expected behavior:

#### usb_board_reset.mk
```usb_board_reset.mk
# open the ttyACM device at 600 baud to signal application reset
RESETFFLASG ?= $(STTY_FLAG) $(PORT) raw ispeed 600 ospeed 600 cs8 -cstopb ignpar eol 255 eof 255
RESET ?= $(PREFLASHER) $(RESETFFLASG)

# open the ttyACM device at 1200 baud to signal bootloader reset
PREFLASHER ?= stty
PREFFLAGS  ?= $(STTY_FLAG) $(PROG_DEV) raw ispeed 1200 ospeed 1200 cs8 -cstopb ignpar eol 255 eof 255
```

Reply:
Nevermind the b600 point -- 600 resets the board, 1200 goes to bootloader, and both work.